### PR TITLE
Add gpt-3.5-turbo comparison

### PR DESCRIPTION
**Pull request by arcticfly**
Allow user to compare the performance of their models against gpt-3.5-turbo.

### Changes
* Index FineTuneTestingDatasetEntry by `modelId` instead of `fineTuneId`
* In general, allow comparison model id to be used in the same way `fineTuneId` was being used
* Kick off testing jobs when new dataset entries are added to a dataset

### CODE REVIEW

**Comment by corbt** [comment]
Looks like maybe we aren't passing through function calls properly?
 
Screen Shot 2023-10-17 at 9 22 56 AM

**Comment by corbt** [comment]
Btw I just had a call with a potential client who said they'd much rather use our API key and get a bill than use their own, over rate-limit concerns. Something to keep in mind I guess.

File: app/src/components/datasets/DatasetContentTabs/Evaluation/AddComparisonModelDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                <Text>
+                  To include this model, add your OpenAI API key on the{" "}
+                  <ChakraLink as={Link} href="/project/settings" target="_blank" color="blue.600">
+                    <Text as="span">project settings</Text>
+                  </ChakraLink>{" "}
+                  page.
+                </Text>
+              ) : (
+                <VStack>
+                  <Text>To confirm this change, please type the model's ID below.</Text>
```
Comment (corbt) [inline app/src/components/datasets/DatasetContentTabs/Evaluation/AddComparisonModelDialog.tsx]: Let's add text like "This will result in 123 calls to the OpenAI API". Right now the dialog does a good job of making this scary but not a great job of giving me the information I need to assess whether I should do it or not.

File: app/src/modelProviders/fine-tuned/getCompletion-2.ts [hunk 2]
```diff
@@ -20,20 +20,21 @@ export async function getCompletion2(
...
   } else {
     return getModalCompletion(fineTune, prunedInput);
   }
 }
 
-async function getOpenaiCompletion(
-  fineTune: FineTune,
+export async function getOpenaiCompletion(
+  projectId: string,
+  modelId: string | null,
```
Comment (corbt) [inline app/src/modelProviders/fine-tuned/getCompletion-2.ts]: Can't we just get this from the `input` parameter?

File: app/src/modelProviders/fine-tuned/getCompletion-2.ts [hunk 3]
```diff
@@ -20,20 +20,21 @@ export async function getCompletion2(
...
-    return getOpenaiCompletion(fineTune, prunedInput);
+    return getOpenaiCompletion(fineTune.projectId, fineTune.openaiModelId, prunedInput);
   } else {
     return getModalCompletion(fineTune, prunedInput);
   }
 }
 
-async function getOpenaiCompletion(
-  fineTune: FineTune,
+export async function getOpenaiCompletion(
```
Comment (corbt) [inline app/src/modelProviders/fine-tuned/getCompletion-2.ts]: Since this function is no longer fine-tune specific I think it should be moved to another file (probably just included in our openai.ts). It's confusing to have it inside fine-tunes if it's just a generic function to get a completion with a given project's API key.

File: app/src/server/tasks/evaluateTestSetEntry.task.ts [hunk 4]
```diff
@@ -9,48 +10,58 @@ import {
...
-      where: { fineTuneId_datasetEntryId: { fineTuneId, datasetEntryId } },
+      where: { modelId_datasetEntryId: { modelId, datasetEntryId } },
     });
 
     if (existingTestEntry?.output && !skipCache) return;
 
     let prunedMessages = existingTestEntry?.prunedInput;
 
-    const stringsToPrune = await getStringsToPrune(fineTune.id);
+    const stringsToPrune = await getStringsToPrune(modelId);
```
Comment (corbt) [inline app/src/server/tasks/evaluateTestSetEntry.task.ts:53]: Maybe we can just drop pruned messages from the test set entries? Not sure we actually need them there?

File: app/src/server/tasks/evaluateTestSetEntry.task.ts [hunk 5]
```diff
@@ -60,7 +71,7 @@ export const evaluateTestSetEntry = defineTask<EvaluateTestSetEntryJob>({
     }
 
     const cacheKey = hashObject({
-      fineTuneId,
+      modelId,
       input: prunedMessages,
```
Comment (corbt) [inline app/src/server/tasks/evaluateTestSetEntry.task.ts]: Unrelated to this PR but we need to add the function_call and functions to this hash as well.

File: app/src/server/tasks/evaluateTestSetEntry.task.ts [hunk 6]
```diff
@@ -9,48 +10,58 @@ import {
...
 
-    const datasetEntry = typedDatasetEntry(rawDatasetEntry);
+    let fineTune;
+    if (!isComparisonModel(modelId)) {
+      fineTune = await prisma.fineTune.findUnique({
+        where: { id: modelId },
+      });
+    }
+
+    const datasetEntry = typedDatasetEntry(omit(rawDatasetEntry, "dataset"));
```
Comment (corbt) [inline app/src/server/tasks/evaluateTestSetEntry.task.ts]: What does this `omit` do?

File: app/src/server/tasks/evaluateTestSetEntry.task.ts (hunk 4 above)
Comment (arcticfly) [reply app/src/server/tasks/evaluateTestSetEntry.task.ts:53]: We definitely do need to prune messages for the test set entries, because some fine-tunes have been trained on pruned messages.

**Comment by arcticfly** [comment]
> Looks like maybe we aren't passing through function calls properly?
> 
> Screen Shot 2023-10-17 at 9 22 56 AM

I'm seeing function call output show up properly. Could you test on another dataset?

File: app/src/utils/baseModels.ts [hunk 7]
```diff
@@ -14,3 +14,13 @@ export const displayBaseModel = (baseModel: BaseModel) => {
...
   }
 };
+
+export const isComparisonModel = (modelId: string) =>
+  ComparisonModel[modelId as keyof typeof ComparisonModel] !== undefined;
+
+export const getComparisonModelName = (comparisonModel: ComparisonModel) => {
+  switch (comparisonModel) {
+    case "GPT_3_5_TURBO":
+      return "gpt-3.5-turbo";
```
Comment (corbt) [inline app/src/utils/baseModels.ts]: let's add the `-0613` here so we track model changes explicitly.

File: app/src/modelProviders/fine-tuned/getCompletion-2.ts (hunk 2 above)
Comment (arcticfly) [reply app/src/modelProviders/fine-tuned/getCompletion-2.ts]: No, because sometimes we need to swap out the model that's passed in the input with another model, and reconstruct the input. We could do this swap everywhere we call getOpenaiCompletion, but I think it's cleaner to do it within the function itself.

File: app/src/server/api/routers/datasets.router.ts [hunk 8]
```diff
@@ -82,10 +87,15 @@ export const datasetsRouter = createTRPCRouter({
...
         where: { id: input.id },
         data: {
-          name: input.name,
+          name: input.updates.name,
+          enabledComparisonModels: input.updates.enabledComparisonModels,
         },
       });
 
+      if (input.updates.enabledComparisonModels) {
+        await startDatasetTestJobs(input.id);
```
Comment (corbt) [inline app/src/server/api/routers/datasets.router.ts:96]: Can we run this in a background job? It took a surprisingly long time for the popup to clear for me.

Comment (arcticfly) [reply app/src/server/api/routers/datasets.router.ts:96]: Agreed, I've noticed the same thing.

File: app/src/server/api/routers/datasetEntries.router.ts [hunk 9]
```diff
@@ -352,8 +358,8 @@ export const datasetEntriesRouter = createTRPCRouter({
       const newEntry = await prisma.datasetEntry.create({
         data: {
           messages: parsedInput ?? prevEntry.messages,
-          functions: parsedInput ?? prevEntry.functions,
-          function_call: parsedInput ?? prevEntry.function_call,
+          functions: prevEntry.functions ?? undefined,
```
Comment (corbt) [inline app/src/server/api/routers/datasetEntries.router.ts]: ouch good catch

**Review by corbt — CHANGES_REQUESTED** [review]
Excited to get this in!

File: app/src/server/tasks/evaluateTestSetEntry.task.ts (hunk 4 above)
Comment (corbt) [reply app/src/server/tasks/evaluateTestSetEntry.task.ts:53]: We need them when we run inference yes. But `getCompletion2` handles that internally. It looks like we aren't using any of `prunedInput`, `prunedInputTokens`, `outputTokens` on the `FineTuneTestingEntry` and could just drop them?

File: app/src/server/api/routers/datasets.router.ts (hunk 8 above)
Comment (corbt) [reply app/src/server/api/routers/datasets.router.ts:96]: Yeah just tried again with a 10k row dataset and we're 2 minutes in and the button is still spinning. Definitely feels broken.

File: app/src/server/api/routers/datasetEntries.router.ts (hunk 9 above)
Comment (arcticfly) [reply app/src/server/api/routers/datasetEntries.router.ts]: Yeah. I'm not sure how many users were impacted by this. I ended up merging this change in a PR last night: https://github.com/OpenPipe/OpenPipe/pull/296

File: app/src/modelProviders/fine-tuned/getCompletion-2.ts (hunk 2 above)
Comment (corbt) [reply app/src/modelProviders/fine-tuned/getCompletion-2.ts]: Let's just make the caller swap the model out when constructing the `input` object in those cases. It's really confusing to have an `input.model` parameter which is required but always ignored.

File: app/src/server/utils/startTestJobs.ts [hunk 10]
```diff
@@ -37,6 +68,6 @@ export const startTestJobs = async (fineTune: FineTune) => {
     skipDuplicates: true,
   });
   for (const entry of datasetEntries) {
-    await evaluateTestSetEntry.enqueue({ fineTuneId: fineTune.id, datasetEntryId: entry.id });
```
Comment (corbt) [inline app/src/server/utils/startTestJobs.ts:40]: We need to refactor the way this function works so it first figures out which outputs haven't been generated yet and then only does the `createMany`/`enqueue` on the ones that are actually missing. That'll make this way faster and less resource intensive.

It'll also provide a far better UX. I added a gpt-3.5 comparison 10 minutes ago for a dataset with 2500 test entries and 2 fine tuned models and it still hasn't started showing any of the 3.5 completions since it's still working its way through the 5000 no-op jobs for the existing fine-tunes.

File: app/src/server/tasks/evaluateTestSetEntry.task.ts (hunk 6 above)
Comment (arcticfly) [reply app/src/server/tasks/evaluateTestSetEntry.task.ts]: Nothing, I removed it. I added it because I was getting a type error and wasn't sure if it was due to the rawDatasetEntry having more fields than typedDatasetEntry expected.

File: app/src/server/api/routers/datasets.router.ts (hunk 8 above)
Comment (arcticfly) [reply app/src/server/api/routers/datasets.router.ts:96]: We're no longer calculating prunedInputTokens before creating these test entries, which dramatically speeds up the process.

File: app/src/server/utils/startTestJobs.ts (hunk 10 above)
Comment (arcticfly) [reply app/src/server/utils/startTestJobs.ts:40]: That's what I'm trying to do on line 35. I suspect a lot of the latency you saw was due to our previous step of calculating input tokens.

File: app/src/components/datasets/DatasetContentTabs/Evaluation/AddComparisonModelDialog.tsx (hunk 1 above)
Comment (arcticfly) [reply app/src/components/datasets/DatasetContentTabs/Evaluation/AddComparisonModelDialog.tsx]: Added better feedback:
Screenshot 2023-10-17 at 12 46 44 PM

File: app/src/server/tasks/evaluateTestSetEntry.task.ts (hunk 5 above)
Comment (arcticfly) [reply app/src/server/tasks/evaluateTestSetEntry.task.ts]: Added

File: app/src/server/tasks/evaluateTestSetEntry.task.ts (hunk 4 above)
Comment (arcticfly) [reply app/src/server/tasks/evaluateTestSetEntry.task.ts:53]: As we discussed, we'll want to start showing cost soon, so we need prunedInputTokens and outputTokens. I think keeping prunedInput also makes sense (though I suppose we could calculate it on the fly) because if the test output is horribly wrong, it would be useful to see what the pruned input looked like.

File: app/src/modelProviders/fine-tuned/getCompletion-2.ts (hunk 2 above)
Comment (arcticfly) [reply app/src/modelProviders/fine-tuned/getCompletion-2.ts]: That works for me.

File: app/scripts/copy-db.sh [hunk 11]
```diff
@@ -19,7 +19,7 @@ parse_connection_string() {
 
 # Function to check whether to dump the prod DB
 should_dump_prod_db() {
-    if [ ! -f "/tmp/openpipe-prod.db" ] || [ "$FORCE_DUMP" = "true" ]; then
+    if [ ! -f "$PROD_DUMP_FILE" ] || [ "$FORCE_DUMP" = "true" ]; then
```
Comment (corbt) [inline app/scripts/copy-db.sh:22]: Where is this variable defined? It should either be a required cli arg or have a default value (or both)

File: app/src/server/tasks/evaluateTestSetEntry.task.ts (hunk 4 above)
Comment (corbt) [reply app/src/server/tasks/evaluateTestSetEntry.task.ts:53]: Yeah let's drop `prunedInput` from the db and calculate it on the fly if necessary.

File: app/scripts/copy-db.sh (hunk 11 above)
Comment (arcticfly) [reply app/scripts/copy-db.sh:22]: It's defined in your .env file. Since most developers won't use this script, I'm okay with this DX, but you're welcome to improve it.

**Review by corbt — APPROVED** [review]
